### DISCUSSION TITLE
Lazy loading relationships on wide tables/models

### DISCUSSION THREAD

**Discussion by drew-acheson** [spatie/laravel-data#334] [Q&A]
I have a data object that has quite a few properties in it's constructor. Is there a way to use lazy loading (on the last property listed here) without having to restate all these properties in the `fromModel()` function? An example of my model is below.

```
class MemberData extends Data
{
    public function __construct(
        public int $PersonID,
        public ?string $FirstName,
        public string $LastName,
        public ?string $DisplayName,
        public ?string $Class,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d')]
        public ?DateTime $ClassChangeDate,
        #[MapName('status')]
        public ?string $STATUS,
        #[Rule(['email'])]
        public ?string $Email,
        public ?string $PhoneHome,
        public ?string $PhoneWork,
        public ?string $Address,
        public ?string $Address2,
        public ?string $City,
        public ?string $StateProv,
        public ?string $AdministrativeArea,
        public ?string $AdministrativeAreaISO,
        public ?string $Country,
        public ?string $PostalCode,
        public ?string $Nationality,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d')]
        public ?DateTime $ChangeOfAddressDate,
        public ?string $Gender,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d')]
        public ?DateTime $BirthDate,
        public ?string $Rating,
        public $RatingEffectiveDate,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d')]
        public ?DateTime $RatingToDate,
        public ?int $RatingRoundsUsed,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d')]
        public ?DateTime $ReceiveDate,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d')]
        public ?DateTime $JoinDate,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d')]
        public ?DateTime $ExpDate,
        public ?string $MemberType,
        public ?string $AvatarURL,
        public int $AvatarUpdatedAt,
        public ?string $Visibility,
        #[MapName('hof')]
        #[WithCast(YesNoBooleanCast::class)]
        public ?bool $HOF,
        #[WithCast(YesNoBooleanCast::class)]
        public ?bool $Deceased,
        public ?string $Notes,
        public ?string $EntryUser,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d H:i:s')]
        public ?DateTime $EntryDate,
        public ?string $UpdateUser,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d H:i:s')]
        public ?DateTime $UpdateDate,
        #[WithCast(YesNoBooleanCast::class)]
        public ?bool $Sync,
        #[WithCast(DateTimeInterfaceCast::class, format: 'Y-m-d H:i:s')]
        public ?DateTime $SyncDate,
        #[DataCollectionOf(LiveResultData::class)]
        public Lazy|DataCollection $LiveResults
    ){}
}
```

**Comment by jbeales** [comment]
I'm new to laravel-data too, but here's what I'm doing in my fromModel method:

```

    public static function fromModel(ProviderAccount $account) {
        $normalizer = new ModelNormalizer();
        $normalized = $normalizer->normalize($account);

        $normalized['managed_by'] = Lazy::create(fn() => UserData::collection($account->managedBy));
        $normalized['listings'] = Lazy::create(fn() => ListingData::collection($account->listings));
        $normalized['ratings'] = Lazy::create(fn() => RatingData::collection($account->ratings));

        return new self(...$normalized);
    }
```

This way I let spatie/laravel-data's model normalizer do most of the work, then set the things I want to be lazy as lazy.

**Reply by drew-acheson** [reply]
Thanks, this is helpful.  I am noticing that for some reason the casts aren't working correctly when I use this method though.  For instance, I get the error `"API\\DAL\\Member\\MemberData::__construct(): Argument #23 ($BirthDate) must be of type ?DateTime, string given, called in /var/www/html/api/DAL/Member/MemberData.php on line 128",` when I implement this below function for `fromModel()`.

```
    public static function fromModel(Member $member): self
    {
        $normalizer = new ModelNormalizer();
        $normalized = $normalizer->normalize($member);
        $normalized['LiveResults'] = Lazy::whenLoaded('LiveResults', $member, fn() => LiveResultData::collection($member->LiveResults));

        return new self(...$normalized);
    }
```

Thanks!

**Reply by jbeales** [reply]
Take a look at your Member model, and make sure BirthDate is in the $casts array.

**Reply by drew-acheson** [reply]
Does it need to be? If I remove the `fromModel()` function and the Lazy property at the end of the constructor and pass the same model data to `MemberData::from()` it works just fine.

**Reply by jbeales** [reply]
I'm not sure, but it helped me. I was having trouble with date casts too, then it started working and I don't 100% know why. I had added a $casts array to my model, but only put a different, boolean, property in it. I also made sure that the WithCast and DateTimeInterfaceCast classes were in the use statement at the top of my Data class.

Edit: I looked at the code: I changed the date from having a DateTime class to a Carbon class.

**Reply by jbeales** [reply]
I wonder if by using the normalizer in fromModel we're skipping over parts of the [data pipeline](https://spatie.be/docs/laravel-data/v2/advanced-usage/pipeline).

**Reply by drew-acheson** [reply]
Yes, I believe you are correct. The documentation for pipes says: 

"When using a magic creation methods, the pipeline is not being used (since you manually overwrite how a data object is constructed). Only when you pass in a request object a minimal version of the pipeline is used to authorize and validate the request."

And while I added the appropriate date fields to the $casts array of the model, I have a custom Cast in my Data object class which is not being implemented and returning me bad data.  Need to figure out how to call the `CastPropertiesDataPipe` pipe so the Data Object class does the casting and not the Model class.

**Reply by drew-acheson** [reply]
I was able to implement the `CastPropertiesDataPipe` in the `fromModel` function.  Here is my final function:

```
    public static function fromModel(Member $member): self
    {
        $pipeline = DataPipeline::create()
            ->into(static::class)
            ->through(CastPropertiesDataPipe::class);

        $pipeline->normalizer(new ModelNormalizer());
        $normalized = $pipeline->using($member)->execute();
        $normalized['live_results'] = Lazy::whenLoaded('liveResults', $member, fn() => LiveResultData::collection($member->liveResults));

        return new self(...$normalized);
    }
```

Thanks for all the help @jbeales

**Reply by jbeales** [reply]
Coming back in 2024 to say that the solution of using the normalizer in the `fromModel()` method doesn't work in v4, because the `ModelNormalizer` now returns a `NormalizedModel` object and not an array. Removing the `fromModel` and putting its logic into [an additional pipe](https://spatie.be/docs/laravel-data/v4/advanced-usage/pipeline#content-extending-the-pipeline-within-your-data-class) for the pipeline is how I'm fixing the problem now, (although I have added validation and am having trouble with that, but I think that's a separate problem).
